language/move-binary-format/src/compatibility.rs

```diff
@@ -156,10 +147,15 @@ impl Compatibility {
             {
                 // if it was public(script), it must remain public(script)
                 // if it was not public(script), it _cannot_ become public(script)
-                old_func.is_entry == new_func.is_entry
+                // if it was private before, it can become public(script)
+                old_func.visibility == Visibility::Private || old_func.is_entry == new_func.is_entry
             } else {
-                // If it was an entry function, it must remain one.
                 // If it was not an entry function, it is allowed to become one.
+                // If it was a entry function, it must remain one if unless we set
+                // `check_private_entry_linking` to false.
+                if old_func.is_entry && !new_func.is_entry {
+                    entry_linking = false;
+                }
```

Comment:
This check is literally the opposite of the boolean condition below, right?  That makes me think something duplicative/redundant is going on here.

Reply:
Good spot. What we actually want to do here is say `entry_linking &&= is_entry_compatible` I believe.